lib/model/service_map.go

```diff
@@ -0,0 +1,98 @@
+// Copyright (C) 2023 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package model
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/syncthing/syncthing/lib/svcutil"
+	"github.com/thejerf/suture/v4"
+)
+
+// A serviceMap is a utility map of arbitrary keys to a suture.Service of
+// some kind, where adding and removing services ensures they are properly
+// started and stopped on the given Supervisor. The serviceMap is itself a
+// suture.Service and should be added to a Supervisor.
+type serviceMap[K comparable, S suture.Service] struct {
+	services   map[K]S
+	tokens     map[K]suture.ServiceToken
+	supervisor *suture.Supervisor
+}
+
+func newServiceMap[K comparable, S suture.Service]() *serviceMap[K, S] {
+	m := &serviceMap[K, S]{
+		services: make(map[K]S),
+		tokens:   make(map[K]suture.ServiceToken),
+	}
+	m.supervisor = suture.New(m.String(), svcutil.SpecWithDebugLogger(l))
+	return m
+}
+
+// Add adds a service to the map, starting it on the supervisor. If there is
+// already a service at the given key, it is removed first.
+func (s *serviceMap[K, S]) Add(k K, v S) {
+	if tok, ok := s.tokens[k]; ok {
+		// There is already a service at this key, remove it first.
```

Comment:
Alternatively, we could call this incorrect usage and panic, since it's likely to be caused by a missing Remove at some point...

Reply:
Or a middle ground: Panic if it's an RC, send a failure event (and thus report) otherwise.